File: c++/10942.cpp
#include <iostream>

using namespace std;

int arr[2001];
int dp[2001][2001];

int main() {
  freopen("example.txt", "r", stdin);

  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int N;
  cin >> N;

  for (int i = 1; i <= N; i++) {
    cin >> arr[i];
  }

  for (int i = 1; i <= N; i++) {
    dp[i][i] = 1;
    if (i + 1 <= N && arr[i] == arr[i + 1]) dp[i][i + 1] = 1;
  }

  for (int diff = 2; diff < N; diff++) {
    for (int i = 1; i + diff <= N; i++) {
      int j = i + diff;
      if (arr[i] == arr[j] && dp[i + 1][j - 1]) dp[i][j] = 1;
    }
  }

  int M;
  cin >> M;

  while (M--) {
    int a, b;
    cin >> a >> b;
    cout << dp[a][b] << '\n';
  }

  return 0;
}

File: js/10942.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .trim()
  .split('\n')
  .map((e) => e.split(' ').map(Number));

const answer = [];

let idx = 0;
const getLine = () => inputs[idx++];

const [N] = getLine();
const arr = [0, ...getLine()];

const dp = Array(N + 1)
  .fill()
  .map(() => Array(N + 1).fill(0));

let [M] = getLine();
for (let i = 1; i <= N; i++) {
  dp[i][i] = 1;
  if (i + 1 <= N) {
    if (arr[i] === arr[i + 1]) dp[i][i + 1] = 1;
  }
}

for (let diff = 2; diff <= N; diff++) {
  for (let i = 1; i + diff <= N; i++) {
    const j = i + diff;
    if (dp[i + 1][j - 1] && arr[i] === arr[j]) dp[i][j] = 1;
  }
}

while (M--) {
  const [i, j] = getLine();
  answer.push(dp[i][j]);
}

console.log(answer.join('\n'));
